### PR TITLE
Use Kill Query for Non-Transaction Query Execution and Update Query Timeout / Cancelled Error Message

### PR DESCRIPTION
**Pull request by harshit-gangal**
## Description

This PR changes the error message returned on the context error.
On context timeout: `ERROR 3024 (HY000): Query execution was interrupted, maximum statement execution time exceeded`
On context cancelled: `ERROR 1317 (70100): Query execution was interrupted`

This PR also executed `kill query` instead of `kill connection` on context error to reduce the connection churn. This is only performed on non-transactional queries.

## Related Issue(s)

- Closes https://github.com/vitessio/vitess/issues/10059

## Checklist

-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

**Review by systay — APPROVED** [review]

File: go/vt/vttablet/tabletserver/connpool/dbconn.go [hunk 1]
```diff
@@ -426,6 +474,49 @@ func (dbc *Conn) KillWithContext(ctx context.Context, reason string, elapsed tim
...
+		log.Warningf("Failed to get conn from dba pool: %v", err)
+		return err
+	}
+	defer killConn.Recycle()
+
+	ch := make(chan error)
+	sql := fmt.Sprintf("kill query %d", dbc.conn.ID())
+	go func() {
+		_, err := killConn.Conn.ExecuteFetch(sql, -1, false)
+		ch <- err
```
Comment (mattlord) [inline go/vt/vttablet/tabletserver/connpool/dbconn.go:517]: Why not close the channel here as we did elsewhere?

File: go/vt/vttablet/tabletserver/connpool/dbconn_test.go [hunk 2]
```diff
@@ -320,66 +320,153 @@ func TestDBKillWithContext(t *testing.T) {
...
+
+	t.Run("context cancel - non-tx exec", func(t *testing.T) {
+		ctx, cancel := context.WithCancel(context.Background())
+		go func() {
+			time.Sleep(10 * time.Millisecond)
+			cancel()
+		}()
+		testContextError(t, ctx, exec,
+			"(errno 1317) (sqlstate 70100): Query execution was interrupted",
+			150*time.Millisecond)
```
Comment (mattlord) [inline go/vt/vttablet/tabletserver/connpool/dbconn_test.go:352]: I think that this test is likely to be flaky in the CI, given these very small time windows. I'd recommend extending all of them at least a bit.

**Review by mattlord — APPROVED** [review]
I had some minor nits/comments/suggestions. I'll approve and you can address those as you feel is best. Thanks!

**Review by GuptaManan100 — APPROVED** [review]

File: go/vt/vttablet/tabletserver/connpool/dbconn.go [hunk 3]
```diff
@@ -195,9 +200,28 @@ func (dbc *Conn) execOnce(ctx context.Context, query string, maxrows int, wantfi
...
+	case errors.Is(ctx.Err(), context.Canceled):
+		errMsg = "(errno 1317) (sqlstate 70100): Query execution was interrupted"
+	default:
+		errMsg = ctx.Err().Error()
+	}
+	if insideTxn {
+		// we can't safely kill a query in a transaction, we need to kill the connection
+		_ = dbc.Kill(errMsg, time.Since(now))
+	} else {
+		_ = dbc.KillQuery(errMsg, time.Since(now))
```
[the remark this answers is not in the document]
Comment (harshit-gangal) [reply go/vt/vttablet/tabletserver/connpool/dbconn.go:218]: If for any reason `kill` fails, we log the error message. Kill itself is not directly in the query execution path. But, it kills any other running query, which will receive the error message if the kill succeeds.

File: go/vt/vttablet/tabletserver/connpool/dbconn.go (hunk 1 above)
Comment (harshit-gangal) [reply go/vt/vttablet/tabletserver/connpool/dbconn.go:517]: I added it in `exec` and `streamexec`, there is no reason we cannot add it here.
Added now.

File: go/vt/vttablet/tabletserver/connpool/dbconn_test.go (hunk 2 above)
Comment (harshit-gangal) [reply go/vt/vttablet/tabletserver/connpool/dbconn_test.go:352]: I have run these tests multiple times and have not observed any flakiness. I will increase them if they come out as flaky.

**Comment by shlomi-noach** [comment]
LGTM

**Pull request merged**